gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/service/token/impl/TokenServiceImpl.java

```diff
@@ -126,11 +126,17 @@ public Maybe<Token> getRefreshToken(String refreshToken, Client client) {
     }
 
     @Override
-    public Single<Token> introspect(String token) {
-        return introspectionTokenService.introspect(token, false)
-                .map(this::convertAccessToken);
+    public Single<Token> introspect(Token token) {
+        if(token instanceof RefreshToken){
+            return introspectionTokenFacade.introspectRefreshToken(token.getValue())
+                    .map(this::convertRefreshToken);
+        } else {
+            return introspectionTokenFacade.introspectAccessToken(token.getValue())
+                    .map(this::convertAccessToken);
+        }
```

Comment:
**discussion**: According to the spec https://datatracker.ietf.org/doc/html/rfc7662#section-2.1

> token_type_hint
      OPTIONAL.  A hint about the type of the token submitted for
      introspection.  The protected resource MAY pass this parameter to
      help the authorization server optimize the token lookup.  If the
      server is unable to locate the token using the given hint, it MUST
      extend its search across all of its supported token types.  An
      authorization server MAY ignore this parameter, particularly if it
      is able to detect the token type automatically.  Values for this
      field are defined in the "OAuth Token Type Hints" registry defined
      in OAuth Token Revocation [[RFC7009](https://datatracker.ietf.org/doc/html/rfc7009)].

So IMO when token_type_hint is missing we have to test access_token and then refresh_token if access_token is unkown. If token_type_hint is provided we have to test first the type but fallback to the other one if the one provided in parameter is unknown... :thinking: wdyt ?

Reply:
yeap, you're right. Ill apply the change
(I need to read rfc more carefully next time)